### PR TITLE
Fix `MetaData.get_section` handling for "source" & "outputs"

### LINKED ISSUE #5111: `source` lists do not allow `path` in conda-build 3.28.x
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

Seen in https://github.com/conda/conda-standalone/pull/36.

This type of `source` metadata:

```
source:
  - path: ../

  - url: https://github.com/conda/conda/archive/{{ conda_version }}.tar.gz
    sha256: 9276686c8a6ee536dc451cc6557685724fe275a44949ac4f741066fd23cdc7b4
    folder: conda_src
    patches:
      - ../src/conda_patches/0001-Rename-and-replace-entrypoint-stub-exe.patch
      - ../src/conda_patches/0002-Manipulate-PATH-directly-instead-of-_call_ing-conda.patch
      - ../src/conda_patches/0003-Return-unknown-module-in-deprecations.patch

  - url: https://github.com/conda/constructor/archive/{{ constructor_version }}.tar.gz  # [win]
    sha256: 787ffd85e9414bdf70fe531f01eab3987a040e3f6a6ac3a01409f4d332f7de9e  # [win]
    folder: constructor_src  # [win]
```

results in this traceback:

```
WARNING: No index specified in get_value('source/path'). Assuming index 0.

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda/exception_handler.py", line 17, in __call__
        return func(*args, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda/cli/main.py", line 83, in main_subshell
        exit_code = do_call(args, parser)
                    ^^^^^^^^^^^^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda/cli/conda_argparse.py", line 140, in do_call
        result = plugin_subcommand.action(getattr(args, "_args", args))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/plugin.py", line 10, in build
        execute(*args, **kwargs)
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/cli/main_build.py", line 572, in execute
        outputs = api.build(
                  ^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/api.py", line 2[54](https://github.com/conda/conda-standalone/actions/runs/7181452098/job/19555878864#step:5:55), in build
        return build_tree(
               ^^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/build.py", line 3793, in build_tree
        packages_from_this = build(
                             ^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/build.py", line 2474, in build
        output_metas = expand_outputs([(m, need_source_download, need_reparse_in_env)])
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/render.py", line 929, in expand_outputs
        for output_dict, m in deepcopy(_m).get_output_metadata_set(
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/metadata.py", line 2[58](https://github.com/conda/conda-standalone/actions/runs/7181452098/job/19555878864#step:5:59)6, in get_output_metadata_set
        conda_packages = finalize_outputs_pass(
                         ^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/metadata.py", line 949, in finalize_outputs_pass
        fm = finalize_metadata(
             ^^^^^^^^^^^^^^^^^^
      File "/usr/share/miniconda/lib/python3.11/site-packages/conda_build/render.py", line [72](https://github.com/conda/conda-standalone/actions/runs/7181452098/job/19555878864#step:5:73)6, in finalize_metadata
        m.meta["source"]["path"] = normpath(join(m.path, source_path))
        ~~~~~~~~~~~~~~~~^^^^^^^^
    TypeError: list indices must be integers or slices, not str
```

Reverting to conda-build=3.27 fixes the issue.

### Conda Info

```shell
active environment : base
    active env location : /usr/share/miniconda
            shell level : 1
       user config file : /home/runner/.condarc
 populated config files : /home/runner/.condarc
          conda version : 23.10.0
    conda-build version : 3.28.1
         python version : 3.11.5.final.0
       virtual packages : __archspec=1=zen2
                          __glibc=2.35=0
                          __linux=6.2.0=0
                          __unix=0=0
       base environment : /usr/share/miniconda  (writable)
      conda av data dir : /usr/share/miniconda/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/runner/conda_pkgs_dir
       envs directories : /usr/share/miniconda/envs
                          /home/runner/.conda/envs
               platform : linux-64
             user-agent : conda/23.10.0 requests/2.31.0 CPython/3.11.5 Linux/6.2.0-1018-azure ubuntu/22.04.3 glibc/2.35 solver/libmamba conda-libmamba-solver/23.11.1 libmambapy/1.5.3
                UID:GID : 1001:127
             netrc file : None
           offline mode : False
```

### Conda Config

```shell
N/A
```

### Conda list

```shell
See https://github.com/conda/conda-standalone/actions/runs/7181452098/job/19555878864
```

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by kenodegard**
### Description

The recipe format supports 9 fields, all of which are expected to be `dict` except "source" and "outputs" which can be either a single `dict` or a `list` of `dict`.

Previously this distinction had to be uniquely handled throughout out the code, e.g.:

https://github.com/conda/conda-build/blob/061063f10c51162707c7c5253daa39b2752fcfcc/conda_build/build.py#L1567-L1569

It instead makes more sense for the `MetaData.get_section` to auto handle this sanitization for us and to always return a `list` of `dict` for "source" and "outputs".

Resolves #5111 

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

**Review by jezdez — APPROVED** [review]

File: conda_build/metadata.py [hunk 1]
```diff
@@ -1358,24 +1388,26 @@ def get_value(self, name, default=None, autotype=True):
...
 
         section_data = self.get_section(section)
-        if isinstance(section_data, dict):
-            assert (
-                not index
-            ), f"Got non-zero index ({index}), but section {section} is not a list."
+        if isinstance(section_data, dict) and index:
+            raise ValueError(
+                f"Got non-zero index ({index}), but section {section} is not a list."
+            )
```
[the remark this answers is not in the document]
Comment (marcelotrevisani) [reply conda_build/metadata.py:1410]: I would change ``dict`` by ``Mapping`` from ``abc.collections``

File: conda_build/metadata.py [hunk 2]
```diff
@@ -1358,24 +1388,26 @@ def get_value(self, name, default=None, autotype=True):
...
         # get correct default
         if autotype and default is None and FIELDS.get(section, {}).get(key):
             default = FIELDS[section][key]()
 
         section_data = self.get_section(section)
-        if isinstance(section_data, dict):
-            assert (
-                not index
-            ), f"Got non-zero index ({index}), but section {section} is not a list."
+        if isinstance(section_data, dict) and index:
```
Comment (marcelotrevisani) [inline conda_build/metadata.py:1407]:
```suggestion
        if isinstance(section_data, Mapping) and index:
```

File: conda_build/metadata.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaimergp) [reply conda_build/metadata.py:1410]: We can raise `AssertionError` too, if needed.

File: news/5112-fix-multiple-sources [hunk 3]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* Update `conda_build.metadata.MetaData.get_section` to always return lists for "source" and "outputs". (#5112)
+
+### Bug fixes
+
+* Fix finalizing recipes with multiple sources. (#5112)
```
Comment (jaimergp) [inline news/5112-fix-multiple-sources:7]:
```suggestion
* Fix finalizing recipes with multiple sources. (#5111 via #5112)
```

File: conda_build/metadata.py [hunk 4]
```diff
@@ -1358,24 +1388,26 @@ def get_value(self, name, default=None, autotype=True):
...
         """
         names = name.split("/")
-        assert len(names) in (2, 3), "Bad field name: " + name
         if len(names) == 2:
             section, key = names
             index = None
         elif len(names) == 3:
             section, index, key = names
-            assert section == "source", "Section is not a list: " + section
+            if section not in OPTIONALLY_ITERABLE_FIELDS:
```
Comment (jaimergp) [inline conda_build/metadata.py:1396]: Unfortunate name, given how dicts are also iterable, technically speaking. Not that we should change it, but I wonder if `MAYBE_LIST_OF_DICTS_FIELDS` (or similar) would have been better.

File: conda_build/metadata.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kenodegard) [reply conda_build/metadata.py:1410]: > will the change of exception type be a problem?

very possible, happy to revert and tackle removing assertions separately

[the remark this answers is not in the document]
Comment (kenodegard) [reply conda_build/metadata.py]: reverted in https://github.com/conda/conda-build/pull/5112/commits/80b5ade810383f4a0339e1913d634ab233b73840

**Review by beeankha — APPROVED** [review]

File: conda_build/metadata.py [hunk 5]
```diff
@@ -1316,9 +1317,11 @@ def parse_until_resolved(
     @classmethod
     def fromstring(cls, metadata, config=None, variant=None):
         m = super().__new__(cls)
-        if not config:
-            config = Config()
-        m.meta = parse(metadata, config=config, path="", variant=variant)
+        m.path = ""
+        m._meta_path = ""
+        m.requirements_path = ""
```
Comment (kenodegard) [inline conda_build/metadata.py:1322]: initialization missing that was discovered by the newly added unit test

Comment (jezdez) [reply conda_build/metadata.py:1322]: Huh!

**Review by jezdez — APPROVED** [review]

File: conda_build/metadata.py [hunk 6]
```diff
@@ -622,6 +622,7 @@ def parse(data, config, path=None):
         "prelink_message": None,
         "readme": None,
     },
+    "extra": {},
```
Comment (mbargull) [inline conda_build/metadata.py:625]: `extra` could be anything else (scalar or sequence, not just mapping) acc. to https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#extra-section .
This broke an assumption in `conda-smithy` at least: https://github.com/conda-forge/conda-smithy/pull/1816 .

So, technically a breaking change; but I'd be okay with requiring a mapping here.
This needs an update to the docs, though.

Comment (mbargull) [reply conda_build/metadata.py:625]: Opened gh-5125 to track this.

**Pull request merged**